test/policy_integration/dnspolicy_controller_test.go

```diff
@@ -137,6 +137,55 @@ var _ = Describe("DNSPolicy", func() {
 			}, TestTimeoutMedium, time.Second).Should(Succeed())
 		})
 
+		It("should not process gateway with inconsistent addresses", func() {
+			// build invalid gateway
+			gateway = testutil.NewGatewayBuilder("test-gateway", gatewayClass.Name, testNamespace).
+				WithHTTPListener(TestListenerNameOne, TestHostOne).Gateway
+			Expect(k8sClient.Create(ctx, gateway)).To(BeNil())
+
+			// ensure gateway exists
+			Eventually(func() error {
+				return k8sClient.Get(ctx, client.ObjectKey{Name: gateway.Name, Namespace: gateway.Namespace}, gateway)
+			}, TestTimeoutMedium, TestRetryIntervalMedium).ShouldNot(HaveOccurred())
+
+			// invalidate gateway by setting inconsistent addresses
+			Eventually(func() error {
+				gateway.Status.Addresses = []gatewayv1beta1.GatewayAddress{
+					{
+						Type:  testutil.Pointer(gatewayv1beta1.IPAddressType),
+						Value: TestClusterNameOne + "/" + TestIPAddressOne,
+					},
+					{
+						Type:  testutil.Pointer(utils.MultiClusterIPAddressType),
+						Value: TestIPAddressTwo,
+					},
+				}
+				gateway.Status.Listeners = []gatewayv1beta1.ListenerStatus{
+					{
+						Name:           TestClusterNameOne + "." + TestListenerNameOne,
+						SupportedKinds: []gatewayv1beta1.RouteGroupKind{},
+						AttachedRoutes: 1,
+						Conditions:     []metav1.Condition{},
+					},
+					{
+						Name:           TestListenerNameOne,
+						SupportedKinds: []gatewayv1beta1.RouteGroupKind{},
+						AttachedRoutes: 1,
+						Conditions:     []metav1.Condition{},
+					},
+				}
+				return k8sClient.Status().Update(ctx, gateway)
+			}, TestTimeoutMedium, TestRetryIntervalMedium).ShouldNot(HaveOccurred())
+
+			// expect no dns records
+			Consistently(func() []v1alpha1.DNSRecord {
+				dnsRecords := v1alpha1.DNSRecordList{}
+				err := k8sClient.List(ctx, &dnsRecords, client.InNamespace(dnsPolicy.GetNamespace()))
+				Expect(err).ToNot(HaveOccurred())
+				return dnsRecords.Items
+			}, time.Second*15, time.Second).Should(BeEmpty())
+		})
```

Comment:
A better, or at least additional, check here would probably be to check the dnspolicy status is as expected see https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/test/policy_integration/dnspolicy_controller_test.go#L100-L110. It should be false with an error in the message.

The `Consistently` check for no records is probably fine, but we need to be careful adding many of these as they do slow down the whole test suite.

Reply:
Added an extra check, thanks! 
It feels strange to check for the absence of something with `Eventually` :thinking:  But yes, we should not use it often.